File: vuln-sandbox/python/cwe-347/jwt-none-algorithm.py
# VULNERABLE: CryptoJwtNoneAlgorithm — JWT decoded with verify_signature=False and algorithm='none'
# Rule: CryptoJwtNoneAlgorithmTemplate | CWE-347 | Severity: CRITICAL

import jwt
from flask import Flask, request, jsonify

app = Flask(__name__)


@app.route('/api/auth/verify', methods=['POST'])
def verify_token():
    """Verify a JWT and return the decoded claims."""
    data = request.get_json()
    token = data.get('token', '')

    # <-- VULNERABLE: signature verification disabled — any token is accepted as valid
    # An attacker can craft a token with algorithm='none' and arbitrary claims (e.g., admin: true)
    # and it will be accepted without any cryptographic check
    decoded = jwt.decode(
        token,
        options={"verify_signature": False},
        algorithms=['none'],
    )

    return jsonify({
        'user_id': decoded.get('sub'),
        'role': decoded.get('role'),
        'valid': True,
    })


if __name__ == '__main__':
    app.run()

File: vuln-sandbox/node/cwe-347/jwt-none-algorithm.js
// VULNERABLE: jwt-none-algorithm — JWT verification allows 'none' algorithm
// Rule: CryptoJwtNoneAlgorithm | CWE-347 | Severity: CRITICAL

const express = require('express');
const jwt = require('jsonwebtoken');

const app = express();
app.use(express.json());

const SECRET = process.env.JWT_SECRET || 'supersecret';

app.get('/protected', (req, res) => {
  const token = req.headers.authorization?.split(' ')[1];

  // VULNERABLE: allowing 'none' algorithm lets attackers forge tokens without a signature
  const decoded = jwt.verify(token, SECRET, { algorithms: ['none'] });

  res.json({ user: decoded });
});

app.listen(3000);
